### DISCUSSION TITLE
Initialize the library in AppDelegate and set the delegate method

### DISCUSSION THREAD

**Discussion by asa08** [pointfreeco/swift-composable-architecture#1359] [Q&A] [answered]
I would like to create an AppDelegateCore and use TCA for AppDelegate processing.

```
enum AppDelegateCore {}

// MARK: - State
extension AppDelegateCore {
    struct State: Equatable {
    }
}

// MARK: - Action
extension AppDelegateCore {
    enum Action: Equatable {
        case didFinishLaunching
    }
}

// MARK: - Environment
extension AppDelegateCore {
    struct Environment {
    }
}

// MARK: - Reducer
extension AppDelegateCore {
    static let reducer = Reducer<State, Action, Environment>.combine(
        .init { _, action, environment in
            switch action {
                case .didFinishLaunching:
                    // TODO: Library initialization
                    return .none
            }
        }
    )
}
```

I tried to initialize the library with didFinishLaunchingWithOptions, but there is a delegate method in the initialization process and I am wondering how to implement it.

Please advise me if you have a better way.

Thank you.

**Comment by acosmicflamingo** [comment, marked as answer]
Brandon and Stephen actually have an open-source project in the App Store called isowords that may help you here: https://github.com/pointfreeco/isowords/tree/main/Sources/AppFeature. It's certainly no "Hello World" app, but it's something you may find useful :)

**Reply by asa08** [reply]
Thanks for the reply.

Yes, I have read ioswords many times, but I have not found any sample that implements delegate on initialization of an external library.

For example, adjust's SDK requires the implementation of a delegate method like this.
```
adjustConfig?.delegate = self
```
This callback must be received somewhere. Here is an example of an AppDelegate.
```
extension AppDelegate: AdjustDelegate {
}
```

What do you do in such situations?

**Reply by asa08** [reply]
Oh, sorry.
I found a very helpful part: you are receiving the delegate callback in the client.

https://github.com/pointfreeco/isowords/blob/f132bd03cf557d99e01fec118b4224b078e9f4e8/Sources/ComposableUserNotifications/Live.swift
